femtools/Particles.F90

```diff
@@ -700,375 +746,362 @@ subroutine update_particle_subgroup_fields(state, ele, lcoords, p_list)
+                  "/attributes/attribute/python_fields/store_old_attribute")
+             attribute_size(3) = ndiagnostic + nprescribed + nprognostic
+           end if
+         end do
+
+         call get_option(trim(subgroup_path) // "/name", name)
+         call checkpoint_particles_subgroup(state, prefix, lpostfix, cp_no, particle_lists(list_counter), &
+              attribute_size, name, subgroup_path, subgroup_path_name, output_comm)
+         list_counter = list_counter + 1
        end do
-    end do
+     end do
+
+     if (present(number_of_partitions)) then
+       call mpi_comm_free(output_comm, ierr)
+       call mpi_group_free(output_group, ierr)
+     end if
 
   end subroutine checkpoint_particles_loop
 
-  subroutine checkpoint_particles_subgroup(state,prefix,lpostfix,cp_no,particle_list,attribute_size,name, subgroup_path_name)
+  subroutine checkpoint_particles_subgroup(state, prefix, lpostfix, cp_no, particle_list, attribute_size, name, subgroup_path, subgroup_path_name, output_comm)
     !!<Checkpoint Particles
 
     type(state_type), dimension(:), intent(in) :: state
     character(len = *), intent(in) :: prefix
     character(len = *), intent(in) :: lpostfix
     integer, optional, intent(in) :: cp_no !Checkpoint number of the simulation
     character(len = *), intent(in) :: name
-    character(len=OPTION_PATH_LEN), intent(in) :: subgroup_path_name
+    character(len=OPTION_PATH_LEN), intent(in) :: subgroup_path, subgroup_path_name
 
     type(detector_linked_list), intent(inout) :: particle_list
     integer, dimension(3), intent(in) :: attribute_size
 
-    integer(KIND=MPI_OFFSET_KIND) :: location_to_write, offset
-    
-    type(detector_type), pointer :: node
-    character(len = OPTION_PATH_LEN) :: particles_cp_filename
+    integer, optional, intent(in) :: output_comm !! MPI communicator to use for output/collectives
+
+    character(len=OPTION_PATH_LEN) :: particles_cp_filename
+    character(len=FIELD_NAME_LEN) :: attname
+    integer :: i, j, dim, old_attrib, old_field, commsize, ierr
+    integer(kind=8) :: h5_id, h5_prop, h5_ierror
+    integer(kind=8), dimension(:), allocatable :: npoints
+    real, dimension(:,:), allocatable :: positions, attrib_data, old_attrib_data, old_field_data
+    integer, dimension(:), allocatable :: node_ids
     type(vector_field), pointer :: vfield
+    type(scalar_field), pointer :: sfield
+    type(detector_type), pointer :: node
 
-    integer, ALLOCATABLE, DIMENSION(:) :: status
-    real, dimension(:), allocatable :: buffer
-    
-    integer, save :: fhdet=0
-    integer :: j, ierror
-    integer :: nints, realsize, dimen, num_particles, number_total_columns
+    integer :: comm
+    comm = MPI_COMM_FEMTOOLS
+    if (present(output_comm)) comm = output_comm
 
-    num_particles = particle_list%total_num_det
+    ! we store the number of points per process, so gather them
+    ! this is because h5part attributes must be agreed upon by all,
```

Comment:
Can you clarify a little more why this is necessary? We've had issues with all-to-all communication in the past: when running with large number of cores (multiple thousands), this would be the first thing to fall over. So I've removed a number of point in fluidity where this used to happen. The only places it still happens, I _think_, is actually in detector/particle code (which has never been tested on that many cores). It would be nice if we could avoid it....

Reply:
Good point, I'll try to dig up the documentation where I read this. I'll experiment with different methods -- using the non-collective IO might mean the attribute can only be written by a single process.

Reply:
Okay, so writing the attribute is considered a _metadata-modifying operation_. These must be called collectively to ensure that each process has a consistent view of the dirty (in-memory) metadata cache. At some point, the cache overflows and is written to disk. However, process 0 manages the metadata cache (according to https://www.alcf.anl.gov/files/Parallel_HDF5_1.pdf "Managing the metadata cache"). Indeed, by https://support.hdfgroup.org/HDF5/doc/RM/CollectiveCalls.html

> Because raw data for an attribute is cached locally, all processes must participate in order to guarantee that future `H5Aread` calls return correct results on all processes. 

It may be the case that because we don't need to read the attribute back out, we can get away with a gather alone on the root, and the rest of the processes will just have an incorrect value of the array in their metadata cache.